xla/service/float_normalization.cc

```diff
@@ -564,6 +570,58 @@ Status FloatNormalizationVisitor::Preprocess(HloInstruction* hlo) {
   return OkStatus();
 }
 
+// We must avoid normalizing computations that have non-normalizing users
+// (e.g., all-reduce's computations should not be normalized). If a
+// computation is shared between normalizing and non-normalizing users, we will
+// clone the computation for the non-normalizing users so that it can be left
+// unchanged. This function clones the shared computations and returns the set
+// of non-normalizing computations that must be skipped by the visitor.
+absl::flat_hash_set<HloComputation*>
+CloneComputationsForNonNormalizingInstructions(
+    HloModule* module, std::vector<HloComputation*>& computations_to_visit) {
+  absl::flat_hash_set<HloComputation*> computations_to_skip;
+  // Let us first figure out what users each computation has.
+  struct ComputationUsers {
+    bool user_to_normalize_exists = false;
+    std::vector<HloInstruction*> users_to_skip_normalization;
+  };
+  absl::flat_hash_map<HloComputation*, ComputationUsers> computation_users;
+  for (auto& computation : computations_to_visit) {
+    for (auto* instruction : computation->instructions()) {
+      if (!instruction->has_called_computations()) continue;
+      for (auto& called : instruction->called_computations()) {
+        if (ShouldSkipCalledComputationsForInstruction(instruction)) {
+          computation_users[called].users_to_skip_normalization.push_back(
+              instruction);
+        } else {
+          computation_users[called].user_to_normalize_exists = true;
+        }
+      }
+    }
+  }
+
+  // If a computation is shared, create a clone, and replace it its the users.
+  for (auto [comp, users] : computation_users) {
```

Comment:
Nit: it's quite a bit of mental overhead to follow the code above and to understand what it actually does. WDYT about instead constructing a CallGraph (what passes usually do), and then simply iterating over `GetComputationCallers()` for each instruction? Then we don't need a 2-pass system.

Reply:
If don't mind introducing the extra dependency and overhead, I am all for that.

That said, we still need a little 2 pass system (1 - to figure out if the computation is shared, 2 - to replace the cloned computation in instruction), but it does look a bit better.